server/channels/store/sqlstore/job_store.go

```diff
@@ -48,7 +48,62 @@ func (jss SqlJobStore) Save(job *model.Job) (*model.Job, error) {
 	}
 
 	if _, err = jss.GetMasterX().Exec(queryString, args...); err != nil {
-		return nil, errors.Wrap(err, "failed to save Preference")
+		return nil, errors.Wrap(err, "failed to save Job")
+	}
+
+	return job, nil
+}
+
+func (jss SqlJobStore) SaveOnce(job *model.Job) (*model.Job, error) {
+	jsonData, err := json.Marshal(job.Data)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed marshalling job data")
+	}
+	if jss.IsBinaryParamEnabled() {
+		jsonData = AppendBinaryFlag(jsonData)
+	}
+
+	tx, err := jss.GetMasterX().Beginx()
+	if err != nil {
+		return nil, errors.Wrap(err, "begin_transaction")
+	}
+	defer finalizeTransactionX(tx, &err)
+
+	query, args, err := jss.getQueryBuilder().
+		Select("COUNT(*)").
+		From("Jobs").
+		Where(sq.Eq{
+			"Status": []string{model.JobStatusPending, model.JobStatusInProgress},
+			"Type":   job.Type,
+		}).ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "job_tosql")
+	}
+
+	var count int64
+	err = tx.Get(&count, query, args...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to count pending and in-progress jobs with type=%s", job.Type)
+	}
+
+	if count > 0 {
+		return nil, nil
+	}
```

Comment:
I think this just moves the race again: if the two pods both complete the `SELECT COUNT(*)` query successfully, they will proceed to `INSERT` despite the transaction -- at least at our current isolation levels. I imagine this might work the `SERIALIZABLE` isolation level, but we don't (and can't!) require that of customers.

I'm not sure of a way to do this outside of a UNIQUE constraint on a column that ensures a functioning critical section.

Reply:
`SERIALIZABLE` sounds fine to me actually. Why can't we use that?

Reply:
Oh you mean if certain customers don't allow `SERIALIZABLE` statements to run in their DB envs? That seems unlikely because we already have a store method `(s SqlSystemStore) InsertIfExists` which does this exact same thing in a serializable transaction. Don't believe we have seen any issues with that. And moreover, this is only supposed to run in cloud so it should be fine nevertheless. 

I am going to update the PR assuming we are in agreement. But also because we need to ship this fix ASAP or revert the migration PR before this gets pushed out to cloud.

Reply:
Oh, right! Yes, we can rely on `SERIALIZABLE` within just this transaction. I had been thinking of it as a global setting only which would kill performance.